### PR TITLE
Storage chains sync

### PR DESCRIPTION
**Pull request by arkpar**
When syncing storage chains substrate needs to download block bodies in a special way. Since transactions may be renewed, we need to download indexed data rather than block bodies.

### CODE REVIEW

File: client/api/src/in_mem.rs [hunk 1]
```diff
@@ -572,6 +572,7 @@ impl<Block: BlockT> backend::BlockImportOperation<Block> for BlockImportOperatio
 		&mut self,
 		header: <Block as BlockT>::Header,
 		body: Option<Vec<<Block as BlockT>::Extrinsic>>,
+		_indexed_body: Option<Vec<Vec<u8>>>,
```
[the remark this answers is not in the document]
Comment (arkpar) [reply client/api/src/in_mem.rs:575]: They are not mutually exclusive on the client level. Potentially we might want to import both. They are exclusive on the network level only currently

File: client/db/src/lib.rs [hunk 2]
```diff
@@ -1877,6 +1891,20 @@ fn apply_index_ops<Block: BlockT>(
 	extrinsic_headers.encode()
 }
 
+fn apply_indexed_body<Block: BlockT>(
```
[the remark this answers is not in the document]
Comment (arkpar) [reply client/db/src/lib.rs:1894]: Don't think so. These are quite different

File: client/service/src/builder.rs [hunk 3]
```diff
@@ -973,6 +973,14 @@ pub fn build_network<TBl, TExPool, TImpQu, TCl>(
 		light_client_request_protocol_config,
 	};
 
+	// Force fast sync when storage chain mode is enabled
```
[the remark this answers is not in the document]
Comment (arkpar) [reply client/service/src/builder.rs:976]: Storage chains don't keep block history and therefore can't be synced from genesis in regular fashion. I"ll add a comment.

File: client/network/src/block_request_handler.rs [hunk 4]
```diff
@@ -317,6 +318,14 @@ impl<B: BlockT> BlockRequestHandler<B> {
 						break;
 					}
 				}
+			} else if get_indexed_body {
+				match self.client.block_indexed_body(&BlockId::Hash(hash))? {
```
[the remark this answers is not in the document]
Comment (arkpar) [reply client/network/src/block_request_handler.rs:326]: Introduced an additional field

[the remark this answers is not in the document]
Comment (cheme) [reply client/network/src/block_request_handler.rs:326]: I remember adding field to protobuf schema in the past without issue. I think it is fine as long as it is not a 'required' field (so repeating is fwiu)

**Review by cheme — APPROVED** [review]
Changes looks good (sorry for the 11 days delay)

File: client/db/src/lib.rs [hunk 5]
```diff
@@ -1877,6 +1891,20 @@ fn apply_index_ops<Block: BlockT>(
...
+	transaction: &mut Transaction<DbHash>,
+	body: Vec<Vec<u8>>,
+)  {
```
Comment (andresilva) [inline client/db/src/lib.rs:1897]:
```suggestion
) {
```

**Review by andresilva — APPROVED** [review]

**Pull request merged**